clippy_lints/src/and_then_then_some.rs

```diff
@@ -0,0 +1,164 @@
+    match expr.kind {
+        ExprKind::Closure(Closure {
+            fn_decl: FnDecl {
+                inputs: [Ty { hir_id: arg_id, .. }],
+                ..
+            },
+            body,
+            ..
+        }) => {
+            if let Node::Expr(expr) = cx.tcx.hir_node(body.hir_id) {
+                (peel_closure_body(cx, expr, *arg_id)).map(|body| (cx.tcx.hir().span(*arg_id), body))
+            } else {
+                None
+            }
+        },
+        _ => None,
+    }
+}
+
+fn peel_closure_body<'tcx>(
+    cx: &LateContext<'tcx>,
+    expr: &'tcx Expr<'tcx>,
+    closure_arg_id: HirId,
+) -> Option<&'tcx Expr<'tcx>> {
+    match expr.kind {
+        // it would be nice if we could lift { x; y.a() } into { x; y }.a()
+        ExprKind::Block(
+            Block {
+                stmts: [],
+                expr: Some(wrapped_expr),
+                ..
+            },
+            _,
+        ) => peel_closure_body(cx, wrapped_expr, closure_arg_id),
+        ExprKind::MethodCall(_, pred, [arg], _) | ExprKind::Call(_, [pred, arg]) => {
+            if is_then_some(cx, expr) && is_local_defined_at(cx, arg, closure_arg_id) {
+                Some(pred)
+            } else {
+                None
+            }
+        },
+        _ => None,
+    }
+}
+
+fn is_local_defined_at(cx: &LateContext<'_>, local: &Expr<'_>, arg_hid: HirId) -> bool {
+    match local.kind {
+        ExprKind::Path(QPath::Resolved(
+            _,
+            Path {
+                res: Res::Local(local_hid),
+                ..
+            },
+        )) => {
+            // FIXME: this is the best way i could find to compare if
+            // a local refers to a specific closure argument.
+            //
+            // it breaks if the closure argument has an explicitly declared type,
+            // since the spans only align for TyKind::Infer
+            if let Node::Pat(Pat { span: local_span, .. }) = (cx.tcx.hir_node(*local_hid))
+                && let Node::Ty(Ty { span: arg_span, .. }) = (cx.tcx.hir_node(arg_hid))
+                && local_span == arg_span
+            {
+                true
+            } else {
+                false
+            }
+        },
+        // is not local at all, so definitly isn't a local defined at the given position
+        _ => false,
+    }
+}
+
+fn show_sugg(cx: &LateContext<'_>, span: Span, selfarg: &Expr<'_>, closure_args: Span, predicate: &Expr<'_>) {
+    let mut appl = Applicability::MachineApplicable;
+    let sugg = format!(
+        "{}.filter(|{}| {})",
+        snippet_with_applicability(cx, selfarg.span, "<OPTION>", &mut appl),
+        snippet_with_applicability(cx, closure_args, "<ARGS>", &mut appl),
+        snippet_with_applicability(cx, predicate.span, "<PREDICATE>", &mut appl)
+    );
+    span_lint_and_sugg(
+        cx,
+        AND_THEN_THEN_SOME,
+        span,
+        "use of `and_then` + `then_some` is equivelent to `filter`",
```

Comment:
nit: Quickly checked and didn't see any case where `+` was used over `and`, so I think for consistency let's change this:
```suggestion
        "use of `Option::and_then` and `bool::then_some` is equivelent to `filter`",
```